xscen/io.py

```diff
@@ -481,17 +487,33 @@ def coerce_attrs(attrs):
         for i, (name, var) in enumerate(ds.data_vars.items()):
             logger.debug(f"Writing {name} ({i + 1} of {len(ds.data_vars)}) to {path}")
             dsvar = ds.drop_vars(allvars - {name})
-            dsvar.to_zarr(
-                path,
-                mode="a",
-                encoding={k: v for k, v in (encoding or {}).items() if k in dsvar},
-                **zarr_kwargs,
-            )
+            try:
+                dsvar.to_zarr(
+                    path,
+                    mode="a",
+                    encoding={k: v for k, v in (encoding or {}).items() if k in dsvar},
+                    **zarr_kwargs,
+                )
+            except TimeoutException:
+                if timeout_cleanup:
+                    logger.info(f"Removing incomplete {name}.")
+                    sh.rmtree(path / name)
+                raise
+
     else:
         logger.debug(f"Writing {list(ds.data_vars.keys())} for {filename}.")
-        return ds.to_zarr(
-            filename, compute=compute, mode="a", encoding=encoding, **zarr_kwargs
-        )
+        try:
+            return ds.to_zarr(
+                filename, compute=compute, mode="a", encoding=encoding, **zarr_kwargs
+            )
+        except TimeoutException:
```

Comment:
I don't mind keeping that but it doesn't seem that useful to me, if you use a catalog..
If you follow the order  [not exist_in_catlog, xscen function, save_to_zarr, update_catalog], it doesn't matter that the file is half written, it will be overwritten.

Reply:
Also, we use timeout to get out of dask being frozen. Will it even be able to do the removal or will it keep being frozen ?

Reply:
In my case the usefulness comes from the fact that I am _appending_ to the zarr dataset. I think I am indeed not using the catalog to it's full potential. But my workflow creates the full dataset than writes it to zarr and it skips all variables already existing. This timeout cleanup ensures partially written variables are being rewritten on the second pass.

AFAIU, within the `except TimeoutException:` the fact that dask hangs is not important. The `shutil` calls do not call dask, so the files are removed even if any dask object are still "linked" to them. My usecase looks like:

 ```python
for dataset in catalog:
    with Timeout():
        with dask.Client():
            out = do_things(dataset)
            save_to_zarr(out)
```
Thus, on Timeout files are first removed, then the Client is killed and its object are garbage collected and then we go to the next iteration and we restart the counter and recreate a new client.

Reply:
okok I understand that this way makes sens when you write a lot of indicators in one zarr.